---
title: "Data Exploration and Preparation"
author: 
  - name: Florian J. Auer
    url: https://github.com/florian-j-auer
    affiliation: IT-Infrastructure for Translational Medical Research, University of Augsburg
    affiliation_url: https://www.uni-augsburg.de/de/fakultaet/fai/informatik/prof/misit/
date: "03/30/2022"
output: 
  BiocStyle::html_document:
    css: custom.css
    self_contained: no
    toc: yes
    toc_float: yes
  BiocStyle::pdf_document:
    toc: yes

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=55,
                                     args.newline = TRUE,
                                     arrow = TRUE),
                      tidy=TRUE)
```

```{r navigation, eval=HTML, results='asis', echo=FALSE}
res <- knitr::knit_child('navigation/navigation.Rmd', quiet = TRUE)
cat(res, sep = '\n')
```

# Load libraries

```{r libraries, warning=FALSE, message=FALSE}
library(survival)
library(RColorBrewer)
library(dplyr)
library(gplots)
library(igraph)
library(ndexr)
library(RCX)
```

# Patient information

We applied our methods to a large breast cancer patient dataset that we previously studied and preprocessed [27]. That data is compiled out of 10 public microarray datasets measured on Affymetrix Human Genome HG-U133 Plus 2.0 and HG-U133A arrays. The datasets are available from the Gene Expression Omnibus (GEO) [28] data repository (accession numbers GSE25066, GSE20685, GSE19615, GSE17907, GSE16446, GSE17705, GSE2603, GSE11121, GSE7390, GSE6532). The RMA probe-summary algorithm [29] was used to process each of the datasets, and only samples with metadata on metastasis-free survival were selected and combined together on the basis of HG-U133A array probe names. Quantile normalization was applied over all datasets. In the case of several probes mapping to one gene, only the probe with the highest average value was considered.

The patients were assigned to one of two classes: 393 patients with distant metastasis within the first 5 years and 576 patients without metastasis having the last follow-up between 5 and 10 years. Breast cancer molecular subtypes for the patient samples were predicted in [27] utilizing genefu R-package [30].


The constant of labels_GEO_HG.csv

1) People who had metastatic event during the first 0-5 years (correspond to "1"). 393 patients.

2) People who did not have metastatic event during the first five years and who had the last follow up between 5 and 10 years. No metastatic events at all. This class corresponds to "0". 576 patients.

We retrained the Graph-CNN on 872 patients and generated relevances for **97 test patients**.

**It is not stated which patients those 97 patients are!!!**
Use own information

Load the patient information about the 97 patients:

```{r patientsLoad}
patients = read.csv("data/patient_information.csv", stringsAsFactors=F)
head(patients)
```

The patient data is composed of different subtypes with and without metastatic events:

```{r patientsTable}
table(patients[,c("subtype","met.event")])
```

There is a difference between labeled and predicted metastatic event.
Therefore the patients are filtered for only correctly predicted metastatic events:

```{r patientsClean}
## Check colnames
colnames(patients) = gsub("Patient.ID", "geo_accession", colnames(patients))
colnames(patients) = gsub("Predicted", "predicted", colnames(patients))
# colnames(patients) = gsub("Concordance", "concordance", colnames(patients))

## Only select patients, that are correctly classified
patients = patients[patients$Concordance == 1, ]
patients$Concordance = NULL
patients = patients[order(patients$subtype),]

## differentiate between the two groups
selMetastatic = patients$met.event == 1

head(patients)
```


```{r savePatients, eval=FALSE, include=FALSE}
write.csv(patients, "data/processed/patients.csv", row.names = FALSE)
```


Survival by subtype:

```{r patientsSurvival}
survival <- survfit(Surv(mfs.years, met.event) ~ subtype, data=patients)
print(survival)

colors = brewer.pal(5, "Dark2")
names(colors) = c("Basal", "Her2", "LumA", "LumB", "Normal")

plot(survival, col=colors, lwd=2, mark.time=TRUE, 
     xlab="Years since study entry", ylab="Survival"
)
legend("bottomleft", names(colors), col=colors, lwd=2, bty='n')
```


# Gene expression

http://mypathsem.bioinf.med.uni-goettingen.de/resources/glrp
http://mypathsem.bioinf.med.uni-goettingen.de/fileadmin/mypathsem_resources/WP3/GEO_HG_PPI.csv.zip

Gene expression of the whole data set:

```{r geLoad}
ge_all = read.csv("data/GEO_HG_PPI.csv.gz", stringsAsFactors=F)
```

Get the column names for the genes and the patient ids:

```{r geColumns}
ge_gene_col = "probe"
ge_patient_cols = colnames(ge_all)[colnames(ge_all) != ge_gene_col]
```

Get the mean, standard deviation and boundaries for the 25% and 75% quantile for each gene (probe) based on its expression in all patients.

```{r geStatisticsTest, include=FALSE}
process_ge_with_statistics = file.exists("tmp/ge_with_statistics.csv")
```

```{r geStatistics, eval=!process_ge_with_statistics}
ge_with_statistics = ge_all %>%
  rowwise() %>%
  do({
    curRow = unlist(.[ge_patient_cols])
    result = data.frame(., stringsAsFactors = F)
    result["mean"] = mean(curRow)
    result["stdev"] = sd(curRow)
    quartiles = quantile(curRow, probs = c(0.0, 0.25, 0.4, 0.5, 0.6, 0.75, 1.0))
    result["Q25"] = quartiles['25%']
    result["Q50"] = quartiles['50%']
    result["Q75"] = quartiles['75%']
    result
  })

print(unique(gsub("GSM[0-9]+","GSM*",colnames(ge_with_statistics))))
```

```{r geStatisticsSave, eval=!process_ge_with_statistics, include=FALSE}
write.csv(ge_with_statistics, file = "tmp/ge_with_statistics.csv", row.names = FALSE)
```

```{r geStatisticsLoad, eval=process_ge_with_statistics, echo=FALSE}
ge_with_statistics = read.csv(file = "tmp/ge_with_statistics.csv")
print(unique(gsub("GSM[0-9]+","GSM*",colnames(ge_with_statistics))))
```

```{r geStatistics2}
tmp_no_col = dim(ge_with_statistics)[2]
head(ge_with_statistics[,(tmp_no_col-5):tmp_no_col])
```

This data can be used to calculate the expression level of a gene (LOW, NORMAL or HIGH) using quartile boundaries.

```{r geLevelsTest, include=FALSE}
process_ge_expression_level_by_quantiles = file.exists("tmp/ge_expression_level_by_quantiles.csv")
```

```{r geLevels, eval=!process_ge_expression_level_by_quantiles}
ge_expression_level_by_quantiles = ge_with_statistics %>%
    rowwise() %>%
    do({
      curRow = unlist(.[ge_patient_cols])
      l = unlist(.["Q25"])
      h = unlist(.["Q75"])
      
      curRow[ge_patient_cols] = ifelse(curRow < l, "LOW",
                                    ifelse(curRow > h, "HIGH",
                                           "NORMAL"))
      curRow = c(unlist(.[ge_gene_col]), curRow)
      result = as.data.frame(t(curRow), stringsAsFactors=F)
      result
    })

print(unique(gsub("GSM[0-9]+","GSM*",colnames(ge_expression_level_by_quantiles))))
```

```{r geLevelsSave, eval=!process_ge_expression_level_by_quantiles, include=FALSE}
write.csv(ge_expression_level_by_quantiles, file = "tmp/ge_expression_level_by_quantiles.csv", row.names = FALSE)
```

```{r geLevelsLoad, eval=process_ge_expression_level_by_quantiles, echo=FALSE}
ge_expression_level_by_quantiles = read.csv(file = "tmp/ge_expression_level_by_quantiles.csv")
print(unique(gsub("GSM[0-9]+","GSM*",colnames(ge_expression_level_by_quantiles))))
```

```{r geLevels2}
head(ge_expression_level_by_quantiles[,1:7])
```

Get the gene expression only for the patients and order patients by subtype

```{r geOrder}
ge_patients = ge_all[,patients$geo_accession[order(patients$subtype)]]
rownames(ge_patients) = ge_all$probe
```

Check the gene expression data to be normalized, therefore make a boxplot of the expression data:

```{r geBoxplot}
boxColors = sapply(colnames(ge_patients), function(x){
 subtype = patients$subtype[patients$geo_accession == x]
 return(colors[subtype])
})

boxplot(ge_patients, col=boxColors)
par(xpd=TRUE)
legend("topright", legend = names(colors) , 
    col = colors , bty = "o", pch=20 , pt.cex = 3, cex = 1, horiz = FALSE, inset = c(0.03, 0.1))
```


## Differential gene expression

t-test

```{r ttest}
## perform a t-test for each gene
ttest = apply(ge_patients, 1, function(x){
 res = t.test(x[selMetastatic],
              y=x[!selMetastatic])
 return(res$p.value)
})

## adjust for multiple testing
de = data.frame(
 prope=names(ttest), 
 pvalue=ttest, 
 qvalue = p.adjust(ttest)
)

## re-order by q-value
de = de[order(de$qvalue),]

## select differentially expressed genes
de_genes = de$prope[de$qvalue < 0.01]
print(de[de_genes,])
```

```{r saveDE, eval=FALSE, include=FALSE}
write.csv(de, file = "data/processed/de.csv", row.names = FALSE)
```

## Heatmap visualization

A heatmap is a graphical representation of data where the individual values contained in a matrix are represented as colors. Now that we have a subset of DE genes, we can use their counts to generate a heatmap. We expect DE genes to be able to separate the samples from different groups into different clusters of the dendrograms.

```{r heatmap, tidy=FALSE}
# colors
pal <- brewer.pal(9, "RdYlBu")

# heatmap
heatmap.2(
  as.matrix(ge_patients[de_genes,]) , 
  dist=function(x) {as.dist(1-cor(t(x)))}, 
  scale="row",  
  col=pal, 
  colCol = boxColors,
  ColSideColors = ifelse(patients$met.event == 1, "red", "black"),  
  trace="none", 
  cexRow=1, 
  cexCol=0.5,
  dendrogram = "column"
)

## legend
par(xpd=TRUE)

## non-/metastatic
legend("topleft", 
       legend = c("metastatic", "non-metastatic") , 
       col = c("red", "black") , 
       bty = "o", pch=20 , pt.cex = 3, cex = 1, 
       horiz = FALSE, 
       inset=c(-0.1,0.25)
)

## subtypes
legend("bottomleft", 
       legend = names(colors) , 
       col = colors , 
       bty = "o", pch=20 , pt.cex = 3, cex = 1, 
       horiz = FALSE, 
       inset=c(-0.1,-0.27)
)
```


# Relevance score

Paper supplement:

```{r relevanceLoadOriginal, eval=FALSE, include=FALSE}
ppi_relevance_score = readRDS("data/ppi_relevance_score.data.frame.RDS")
pp_tmp = ppi_relevance_score
tmp_pat = pp_tmp$`Patient ID`
pp_tmp$`Patient ID` = NULL
tmp_gene = colnames(pp_tmp)
ppi_relevance_score = as.data.frame(t(pp_tmp))
colnames(ppi_relevance_score) = tmp_pat

ppi_relevance_score$probe = rownames(ppi_relevance_score)
write.csv(ppi_relevance_score, file = "data/ppi_relevance_score.csv", row.names = FALSE)
```


```{r relevanceLoad}
relevance_score = read.csv("data/ppi_relevance_score.csv")
print(unique(gsub("GSM[0-9]+","GSM*",colnames(relevance_score))))
```

```{r relevanceLoad2}
tmp_no_col = dim(relevance_score)[2]
head(relevance_score[,(tmp_no_col-6):tmp_no_col])
```

# PPI networks

We used the Human Protein Reference Database (HPRD) protein-protein interaction (PPI) network [26] as the molecular network to structure the gene expression data. The database contains protein-protein interaction information based on yeast two-hybrid analysis, in vitro and in vivo methods. The PPI network is an undirected graph with binary interactions between pairs of proteins. The graph is not connected.

[26] Keshava Prasad TS, Goel R, Kandasamy K, Keerthikumar S, Kumar S, Mathivanan S, Telikicherla D, Raju R, Shafreen B, Venugopal A, Balakrishnan L, Marimuthu A, Banerjee S, Somanathan DS, Sebastian A, Rani S, Ray S, Harrys Kishore CJ, Kanth S, Ahmed M, Kashyap MK, Mohmood R, Ramachandra YL, Krishna V, Rahiman BA, Mohan S, Ranganathan P, Ramabadran S, Chaerkady R, Pandey A. Human protein reference database?2009 update. Nucleic Acids Res. 2009; 37:767–72. https://doi.org/10.1093/nar/gkn892.

From website http://hprd.org/
Latest release: Apr 13, 2010


## From publication:
http://mypathsem.bioinf.med.uni-goettingen.de/resources/glrp

http://mypathsem.bioinf.med.uni-goettingen.de/fileadmin/mypathsem_resources/WP3/HPRD_PPI.csv.zip

```{r ppiLoad}
ppi_network_matrix = read.csv("data/HPRD_PPI.csv.gz", check.names = FALSE)
ppi_network_igraph = graph_from_adjacency_matrix(as.matrix(ppi_network_matrix), mode = "undirected")

gorder(ppi_network_igraph)
gsize(ppi_network_igraph)
```

Remove multiple edges and loops from one node to itself from the network.

```{r ppiIgraph}
ppi_network_igraph = simplify(ppi_network_igraph, 
                              remove.multiple = T, 
                              remove.loops = T)

gorder(ppi_network_igraph)
gsize(ppi_network_igraph)
```

Prepare the igraph network for converting to RCX, therefore IDs have to be set for nodes and edges.
The CX convention is, in contrast to R, that the IDs start at 0.

```{r ppiAddIds}
V(ppi_network_igraph)$id = seq(0,gorder(ppi_network_igraph)-1)
E(ppi_network_igraph)$id = seq(0,gsize(ppi_network_igraph)-1)

summary(ppi_network_igraph)
```

For the conversion from igraph to RCX, the vertex attribute containing the node names has to be specified.

```{r ppiToRCX}
ppi_network_rcx = fromIgraph(ppi_network_igraph, nodeName = "name")
ppi_network_rcx$metaData
```

Add some information to the network

```{r rcxNetworkAttributes, tidy=FALSE}
networkAttributes <- createNetworkAttributes(
  name = c("author", 
           "name",
           "description",
           "reference",
           "organism",
           "networkType"),
  value = c("Florian J. Auer",
            "Human Protein Reference Database (HPRD) PPI network",
            paste0('Protein-protein interaction (PPI) network from the ',
                   '<a href="http://hprd.org/" target="_blank">',
                   'Human Protein Reference Database (HPRD)</a> ',
                   'used for training and generating subnetworks'),
            paste0('Chereda, H., Bleckmann, A., Menck, K. et al. ',
                   'Explaining decisions of graph convolutional neural networks: ',
                   'patient-specific molecular subnetworks responsible for ',
                   'metastasis prediction in breast cancer. Genome Med 13, 42 (2021). ',
                   '<a href="https://doi.org/10.1186/s13073-021-00845-7" ',
                   'target="_blank">https://doi.org/10.1186/s13073-021-00845-7</a>'),
            "Homo sapiens",
            "Protein-protein interaction")
)
ppi_network_rcx <- updateNetworkAttributes(ppi_network_rcx, networkAttributes)

ppi_network_rcx$metaData
```

```{r ppi_load, eval=FALSE, include=FALSE}
writeCX(ppi_network_rcx, "tmp/HPRD-PPI.cx")
```

## Upload to NDEx

To make the network available to for further analyses, we can upload the network to the NDEx platform (https://www.ndexbio.org/). 
Of course for this an account is required.

```{r ppiNdexUpload, eval=FALSE}
ndex_con = ndex_connect(username = "florianjauer", password = "****")
ndexHPRDuuid = ndex_create_network(ndex_con, ppi_network_rcx)
```

Until now, the network is only visible to the owner.
To change that, and make it visible to everyone, we have to update this property:

```{r ppiNdexVisibility, eval=FALSE}
ndex_network_set_systemProperties(
  ndex_con, 
  ndexHPRDuuid, 
  visibility = TRUE
)
```

## Load from NDEx:

This network is also available on the NDEx platform as "Human Protein Reference Database (HPRD) PPI network":

https://www.ndexbio.org/viewer/networks/079f4c66-3b77-11ec-b3be-0ac135e8bacf

The R package `ndexr` can be used to download the network from NDEx:

```{r ppiNdex}
ndex_con = ndex_connect()
ndexHPRD = ndex_find_networks(ndex_con, "HPRD AND owner:florianjauer")

print(ndexHPRD[c("name", "owner", "externalId", "nodeCount", "edgeCount")])
```


```{r ppiNdexLoad}
ppi_network_rcx = ndex_get_network(ndex_con, "079f4c66-3b77-11ec-b3be-0ac135e8bacf")
print(ppi_network_rcx$metaData)
```


```{r ppiNdexLoad3}
print(ppi_network_rcx$networkAttributes$name)
```

NDEx added automatically the version to the network attributes.


```{r navigation2, eval=HTML, results='asis', echo=FALSE}
res <- knitr::knit_child(HTML_file, quiet = TRUE)
cat(res, sep = '\n')
```

# Session info
```{r sessionInfo}
sessionInfo()
```




















