##============================================================================##
##                         LEVELS OF EVIDENCE                                 ##
##----------------------------------------------------------------------------##
##                                                                            ##
##	         Set of functions to classify gene-drug interactions          ##
## 	           from GDKD and CIViC into levels of evidence.               ##
##                                                                            ##
##============================================================================##




get_levels_GDKD  = function(druggable_gdkd, cancer_gdkd){
  ## Classifies gene-drug interactions found by match_XXX_GDKD into levels of evidence. 
  ## input: druggable_gdkd is a dataframe generated by function match_XXX_GDKD.
  ## input: cancer_gdkd is a character vector with a cancer type used by GDKD.
  ## output: returns the input druggable_gdkd as a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 
  
  ## Check for empty input
  if (length(druggable_gdkd) == 0 || nrow(as.matrix(druggable_gdkd)) == 0){
    message("The input provided was empty. Returning an empty list.")
    X          = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
    colnames(X) = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1")

    return(list(X,X,X,X,X,X,X,X,X,X))
  }
  
  druggable_gdkd[] = lapply(druggable_gdkd, as.character)
  row.names(druggable_gdkd)  = NULL
  druggable_gdkd = druggable_gdkd[,c(1,2,46,3:45)]

  ## Rows with same cancer
  same_cancer=c()
  for (ck in cancer_gdkd){
    same_cancer = c(which(druggable_gdkd$Disease==ck),same_cancer)
  }
  ## Rows with different cancer
  diff_cancer=setdiff(1:nrow(druggable_gdkd), same_cancer)


  ## Generate variables
  header            = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1")
  levelA1           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelA1) = header
  levelA2           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelA2) = header
  levelA3           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelA3) = header
  levelB1           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelB1) = header
  levelB2           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelB2) = header
  levelB3           = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
  colnames(levelB3) = header
  
  ## A levels: Same cancer rows
  for (c in same_cancer){
    ## A1 (consensus)
    consensus = which(druggable_gdkd[c,]=="consensus")
    if (length(consensus)!=0){
      for (l in consensus){
        levelA1[nrow(levelA1)+1,] = druggable_gdkd[c,c(1:6,(l-3):(l+1))]
        row.names(levelA1)=NULL
      }
    }
   
    ## A2 and A3 (emerging)
    emerging = which(druggable_gdkd[c,]=="emerging")
    if (length(emerging)!=0){
      for (l in emerging){
        levelA2[nrow(levelA2)+1,] = druggable_gdkd[c,c(1:6,(l-3):(l+1))]
        row.names(levelA2)=NULL
      }
    }
  }


  ## B levels: Different cancer rows
  for (d in diff_cancer){
    ## B1 (consensus)
    consensus = which(druggable_gdkd[d,]=="consensus")
    if (length(consensus)!=0){
      for (l in consensus){
        levelB1[nrow(levelB1)+1,] = druggable_gdkd[d,c(1:6,(l-3):(l+1))]
        row.names(levelB1)=NULL
      }
    }
    ## B2 and B3 (emerging)
    emerging = which(druggable_gdkd[d,]=="emerging")
    if (length(emerging)!=0){
      for (l in emerging){
        levelB2[nrow(levelB2)+1,] = druggable_gdkd[d,c(1:6,(l-3):(l+1))]
        row.names(levelB2)=NULL
      }
    }
  }
  
  ## Subset emerging evidence into specific levels
  levelA2a = levelA2[which(levelA2$Status=="late trials"),]
  levelA2b = levelA2[which(levelA2$Status=="early trials"),]
  levelA2c = levelA2[which(levelA2$Status=="case report"),]
  levelA3  = levelA2[which(levelA2$Status=="preclinical"),]
  levelB2a = levelB2[which(levelB2$Status=="late trials"),]
  levelB2b = levelB2[which(levelB2$Status=="early trials"),]
  levelB2c = levelB2[which(levelB2$Status=="case report"),]
  levelB3  = levelB2[which(levelB2$Status=="preclinical"),]

  return(list(levelA1,levelA2a,levelA2b,levelA2c,levelA3,levelB1,levelB2a,levelB2b,levelB2c,levelB3))
}




get_levels_CIVIC  = function(druggable_civic, cancer_civic){
  ## Classifies gene-drug interactions found by match_XXX_CIVIC into levels of evidence. 
  ## input: druggable_civic is a dataframe generated by function match_XXX_CIVIC.
  ## input: cancer_civic is a character vector with a cancer type used by CIVIC.
  ## output: returns the input druggable_civic as a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 
  
  ## Check for empty input
  if (length(druggable_civic) == 0 || nrow(as.matrix(druggable_civic)) == 0){
    message("The input provided was empty. Returning an empty list.")
    X = data.frame(matrix(ncol=11,nrow=0), stringsAsFactors = F)
    colnames(X) = c("Disease","Gene","Patient_variant","Variant","Description","Effect","Association","Therapeutic.context","Status","Evidence","PMID_1")

    return(list(X,X,X,X,X,X,X,X,X,X))
  }

  ## Remove factors
  druggable_civic[]= lapply(druggable_civic, as.character)
  row.names(druggable_civic)=NULL
  
  ## Rows with same cancer
  same_cancer=c()
  for (ck in cancer_civic){
    same_cancer = c(which(druggable_civic$disease==ck),same_cancer)
  }
  ## Rows with different cancer
  other_cancer  = setdiff(1:nrow(druggable_civic),same_cancer)

  ## Generate variables
  levelA1            = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelA2a           = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelA2c           = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelA3            = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelB1            = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelB2a           = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelB2c           = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  levelB3            = data.frame(matrix(ncol=8,nrow=0), stringsAsFactors = F)
  
  ## Subset same cancer rows into A levels
  levelA1  = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="A")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelA2a = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="B")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelA2c = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="C")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelA3  = druggable_civic[intersect(same_cancer,which(druggable_civic$evidence_level=="D")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]

  ## Subset different cancer rows into B levels
  levelB1  = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="A")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelB2a = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="B")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelB2c = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="C")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]
  levelB3  = druggable_civic[intersect(other_cancer,which(druggable_civic$evidence_level=="D")),c("disease","gene","Patient_variant", "variant","clinical_significance","drugs","evidence_level","pubmed_id")]

  ## Homogeneize colnames with GDKD 
  colnames(levelA1)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelA2a) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelA2c) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelA3)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelB1)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelB2a) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelB2c) = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")
  colnames(levelB3)  = c("Disease","Gene","Patient_variant","Variant","Association","Therapeutic.context","Status","PMID_1")



  return(list(levelA1,levelA2a,data.frame(matrix(ncol=8,nrow=0)),levelA2c,levelA3,levelB1,levelB2a,data.frame(matrix(ncol=8,nrow=0)),levelB2c,levelB3))
}


merge_levels = function(levels_gdkd, levels_civic){
  ## Merges the output of get_levels_GDKD and get_levels_CIVIC
  ## input: levels_gdkd is a list generated by function get_levels_GDKD.
  ## input: levels_civic is a list  generated by function get_levels_CIVIC.
  ## output: returns a list with 10 positions. 
  ##         Each position corresponds to a level of evidence, following the order: 
  ##         A1, A2a, A2b, A2c, A3, B1, B2a, B2b, B2c, B3. 

  ## Homogeneize columns of levels_gdkd as in levels_civic
  for (i in 1:10){
    levels_gdkd[[i]]     = levels_gdkd[[i]][,c(1:4,6,7:9,11)] 		    # Put in same order as civic columns
    levels_gdkd[[i]][,4] = paste(levels_gdkd[[i]][,4],levels_gdkd[[i]][,5]) # Paste effect to variant (eg. T45X loss-of-function)
    levels_gdkd[[i]]     = levels_gdkd[[i]][,c(1:4,6:9)]
  }

  ## Merge two lists, position by position
  merged = c()
  for (i in 1:10){
    merged[[i]]           = rbind(levels_gdkd[[i]],levels_civic[[i]])
    rownames(merged[[i]]) = NULL
    merged[[i]]           = merged[[i]][order(merged[[i]][,"Gene"]),] #in each level, order by gene
    rownames(merged[[i]]) = NULL
  }
  
  return(merged)

}



clean_levels = function(A,synonyms,sort_by="levels"){
  ## Creates a clean table with gene-drug interactions from list of levels.
  ## input: A is list generated by functions get_levels_GDKD, get_levels_CIVIC or merge_levels.
  ## input: sort_by is a character indicating how to sort the dataframe: "drug_freq", "genes" or "levels".
  ## output: returns a dataframe with a clean annotation of gene-drug interactions. 

if (is.list(A)){
  A1  = A[[1]]
  A2a = A[[2]]
  A2b = A[[3]]
  A2c = A[[4]]
  A3  = A[[5]]
  B1  = A[[6]]
  B2a = A[[7]]
  B2b = A[[8]]
  B2c = A[[9]]
  B3  = A[[10]]
  A = rbind(A1,B1,A2a,A2b,A2c,B2a,B2b,B2c,A3,B3)
}

## To avoid error if A is empty
 if (nrow(A) == 0){
    B = data.frame(matrix(ncol=9,nrow=0), stringsAsFactors = F)
    colnames(B) = c("Cancer","Gene","Pat Var","Known Var","Predicts","Drugs","Evidence","Ref","level")
    return(B)
 }
 
## Homogeneize disease names
civ_disease = lapply(A$Disease,function(x) {
  index = grep(x,synonyms$civic,ignore.case = TRUE)
  return(as.character(synonyms[index,"Report"]))
})

gdkd_disease = lapply(A$Disease,function(x) {
  index = grep(x,synonyms$knowledge,ignore.case = TRUE)
  return(as.character(synonyms[index,"Report"]))
})

civ_disease = sapply(civ_disease,function(x){as.character(x[1])})
gdkd_disease = sapply(gdkd_disease,function(x){as.character(x[1])})
civ_disease[is.na(civ_disease)] = gdkd_disease[which(is.na(civ_disease))]
A$Disease   = civ_disease


## Shorten "amplification" and "deletion"
A$Patient_variant = gsub(pattern = "amplification",replacement = "ampl.",A$Patient_variant)
A$Patient_variant = gsub(pattern = "deletion",replacement = "del.",A$Patient_variant)
A$Patient_variant = gsub(pattern = "_",replacement = "-",A$Patient_variant)

## Variants & Association to lower case
upper               = grep("[A-Za-z][0-9]+",A$Variant)
if (length(upper) !=0 ){
  lower               = tolower(A[-upper,"Variant"])
  A[-upper,"Variant"] = lower
}
if (length(upper) ==0 ){
  A$Variant = tolower(A$Variant)
}

## Shorten and unify variants
A$Variant       = gsub(pattern = "_",replacement = "-",A$Variant)
A$Variant       = gsub(pattern = "any",replacement = "any variant",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "amplification",replacement = "ampl.",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "deletion",replacement = "del.",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "mutation",replacement = "mut.",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "mut[\\.]*",replacement = "any mut.",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "gain-of-function",replacement = "(GoF)",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "loss-of-function",replacement = "(LoF)",A$Variant, ignore.case=TRUE)
A$Variant       = gsub(pattern = "not applicable",replacement = "",A$Variant, ignore.case=TRUE)
#A$Variant       = gsub(pattern = ",",replacement = ", ",A$Variant, ignore.case=TRUE)
A$Association   = tolower(A$Association)


## Homogeneize evidence of civicDB to knowledgeDB categories
A$Status = gsub(pattern = "A$",replacement = "approved",A$Status)
A$Status = gsub(pattern = "B$",replacement = "clin. trials",A$Status)
A$Status = gsub(pattern = "C$",replacement = "case report",A$Status)
A$Status = gsub(pattern = "D$",replacement = "preclinical",A$Status)

## Remove "_" and ^ from PMID
A$PMID_1 = gsub(pattern = "_",replacement = " ",A$PMID_1)
A$PMID_1 = gsub(pattern = "^",replacement = " ",A$PMID_1, fixed=TRUE)

## Remove non informative columns
A$Effect      = NULL
A$Evidence    = NULL
A$Description = NULL
rownames(A)   = NULL

## Add column names and levels of evidence
colnames(A) = c("Cancer","Gene","Pat Var","Known Var","Predicts","Drugs","Evidence","Ref")
A$level     = c(rep("A1",nrow(A1)),rep("B1",nrow(B1)),rep("A2",nrow(A2a)),rep("A2",nrow(A2b)),rep("A2",nrow(A2c)),rep("B2",nrow(B2a)),rep("B2",nrow(B2b)),rep("B2",nrow(B2c)),rep("A3",nrow(A3)),rep("B3",nrow(B3)))

## Aggregate duplicate rows
A_agg   = aggregate(A, by = list(A$Cancer,A$Gene,A$`Pat Var`,A$Predicts,A$Drugs,A$Evidence),
                               FUN = function(X) paste(unique(X), collapse=", "))[,7:15]                             
A=A_agg


#Reorganize table
A = A[,c(2,3,1,4:9)] #Gene, Pat Var, Cancer, Known Var, Predicts, Drugs, Evidence, Ref, level


#Reorder rows according drugs
if(sort_by=="drug"){
  d = unlist(strsplit(A$Drugs," |,|\\)|\\("))
  d = gsub("^ ","",d)
  d = gsub(" $","",d)
  d = tolower(d)
  for (i in c ("","inhibitors","inhibitor","pathway","+","mabs","tkis","in","mutant","blockade","tumor","tumors","plus","wt","setting","or","c","therapy","hormone","monoclonal")){
    d[which(d==i)] = NA
  }
  d = na.omit(d)
  drug_order=c()
  for (drug in names(sort(table(d),decreasing = TRUE))){
    index = grep(drug,A[,6],ignore.case = T)
    index = index[!(index %in% drug_order)]
    drug_order = c(drug_order,index)
  }
  A = A[drug_order,]
}

#Reorder rows according genes
if(sort_by=="genes"){
  A = A[order(A$Gene),]

}

## Remode duplicate rows
A=unique(A)

return(A)
}


summary_actionable_genes = function(A,maf=data.frame(),cnv=data.frame(),druggableTARGET=data.frame()){
  ## Creates a summary table of actionable genes and genomic quality, of provided by user.
  ## input: A is datafrae generated by clean_levels()
  ## input: maf is a data frame
  ## input: cnv is a 
  ## input: druggableTARGET is a dataframe generated by function match_TARGET_MERIC()
  ## output: returns two dataframes (SNVs and CNVs) with a summary of actionable genes (gene, variant, level of evidence, genomic quality)
 tab_summary_snvs=data.frame()
 tab_summary_cnvs=data.frame()

if (nrow(A)!=0|| nrow(druggableTARGET)!=0){
  tab_summary           <- A[,c("Gene","Pat Var","level")]
  tab_summary$level.col <- tab_summary$level
  if (nrow(druggableTARGET)>0){
     B            <- druggableTARGET[,c("Gene","Patient_variant")]
     B$Patient_variant <- gsub(pattern = "amplification",replacement = "ampl.", B$Patient_variant)
     B$Patient_variant <- gsub(pattern = "deletion",replacement = "del.",B$Patient_variant)
     B$level      <- "unknown"
     B$level.col  <- "unknown"
     colnames(B)  <- c("Gene","Pat Var","level","level.col")
     B            <- B[!(B$Gene %in% tab_summary$Gene),]
     tab_summary  <- rbind(tab_summary,B)
  }
  tab_summary$`Pat Var` <- sapply(tab_summary$`Pat Var` , function(x) gsub("NEW", "", x))
  tab_summary$`Pat Var` <- sapply(tab_summary$`Pat Var` , function(x) gsub("([A-Z]) ([A-Z])", "\\1, \\2", x))
  tab_summary$`Pat Var` <- sapply(tab_summary$`Pat Var` , function(x) gsub(" ", "", x))
  tab_summary           <- aggregate(tab_summary[,c("Gene","Pat Var","level.col")],by=list(name=tab_summary$Gene),FUN = function(X) paste(unique(X), collapse=","))[,-1]

  ## SNVs

  if(!is.null(maf)){
    if (nrow(maf)!=0){
      snvs <- tab_summary$Gene %in% maf[,1]
      i <- na.omit(match(tab_summary$Gene,maf[,1]))
      tab_summary_snvs <- cbind(tab_summary[snvs,],maf[i,setdiff(1:ncol(maf),1:3)])
      twice <- grep(",",tab_summary_snvs$`Pat Var`)
      if (length(twice)!=0){
        for (n in 1:length(twice)){
          vars=c()
          vars=strsplit(tab_summary_snvs[twice[n],"Pat Var"],",")[[1]]
          vars=unique(vars)
          if(grepl("ampl\\.|del\\.",paste(vars,collapse=" "))){
            tab_summary_snvs[twice[n],"Pat Var"]=vars[grep("ampl\\.|del\\.",vars,invert=T)]
          }
          else{
            tab_summary_snvs[twice[n],"Pat Var"]=paste(vars,collapse=", ")
          }
        } 
      }
    }
    colnames(tab_summary_snvs) <- c("Gene","Patient's Variant","Level of Evidence",colnames(maf)[setdiff(1:ncol(maf),1:3)])
    #colnames(tab_summary_snvs) <- paste("\\textbf{",colnames(tab_summary_snvs),"}")

  }


 ## CNVs

  tab_summary_cnvs <- tab_summary[grep("ampl\\.|del\\.",tab_summary$`Pat Var`),]
  if (!is.null(tab_summary_cnvs) & nrow(tab_summary_cnvs)!=0 ){
    j <- na.omit(match(tab_summary_cnvs$Gene,cnv[,1]))
    tab_summary_cnvs <- cbind(tab_summary_cnvs,cnv[j,setdiff(1:ncol(cnv),1:2)])
    twice <- grep(",",tab_summary_cnvs$`Pat Var`)
    for (t in twice){
      vars=strsplit(tab_summary_cnvs[t,"Pat Var"],", ")[[1]]
      tab_summary_cnvs[t,"Pat Var"]=vars[grep("ampl\\.|del\\.",vars)]
    }

    colnames(cnv) = gsub("\\^|_"," ",colnames(cnv))
    colnames(tab_summary_cnvs) <- c("Gene","Patient's Variant","Level of Evidence",colnames(cnv)[setdiff(1:ncol(cnv),1:2)])
    #colnames(tab_summary_cnvs) <- paste("\\textbf{",colnames(tab_summary_cnvs),"}")
  }

}
return(list(tab_summary_snvs,tab_summary_cnvs))
}