<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-03-30" />

<title>RCX - R implementation of the Cytoscape Exchange (CX) format</title>

<script src="RCX_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="RCX_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="RCX_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="RCX_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="RCX_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="RCX_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="RCX_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="RCX_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="RCX_files/navigation-1.1/tabsets.js"></script>
<link href="RCX_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="RCX_files/highlightjs-9.12.0/highlight.js"></script>
<script src="RCX_files/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="RCX_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="RCX_files/datatables-binding-0.21/datatables.js"></script>
<link href="RCX_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="RCX_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="RCX_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="RCX_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="RCX_files/crosstalk-1.1.1/js/crosstalk.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="/home/flockowak/R/x86_64-pc-linux-gnu-library/4.0/BiocStyle/resources/html/bioconductor.css" type="text/css" />
<link rel="stylesheet" href="custom.css" type="text/css" />
<link rel="stylesheet" href="bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {

}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 246px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">RCX - R implementation of the Cytoscape Exchange (CX) format</h1>
<p class="author-name">Florian J. Auer<span class="affil-mark">1</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>IT-Infrastructure for Translational Medical Research, University of Augsburg</p>
<h4 class="date">03/30/2022</h4>

</div>


<ul>
<li><a href="Overview.html">Overview</a> <em>A short introduction</em>
<ul>
<li><a href="SoftwareRequirements.html">Software requirements</a> <em>Installation of required packages</em></li>
<li><a href="Ndexr.html">ndexr</a> <em>ndexr - interface with the network data exchange (NDEx)</em></li>
<li><a href="RCX.html">RCX</a> <em>RCX - R implementation of the Cytoscape Exchange (CX) format</em></li>
</ul></li>
<li><a href="DataExplorationAndPreparation.html">Data Exploration and Preparation</a> <em>Loading the source data, explore the content and prepare the data for its usage</em></li>
<li><a href="DataIntegration.html">Data Integration</a> <em>Combine the network data with patient information, gene expression and relevance score</em></li>
<li><a href="Visualization.html">Visualization</a> <em>Explore different ways of visualizing the networks</em>
<ul>
<li><a href="MetaRelSubNetVis.html">MetaRelSubNetVis</a> <em>Comparative web-based visualization</em></li>
<li><a href="NDExEdit.html">NDExEdit</a> <em>Web-based data-dependant visualization with NDExEdit</em></li>
<li><a href="RCXvis.html">RCX based</a> <em>R-based visualization with the RCX package</em></li>
<li><a href="CytoscapeAndRCy3.html">Cytoscape and RCy3</a> <em>Using the RCy3 package to remotly control Cytoscape</em></li>
</ul></li>
</ul>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Networks are a powerful and flexible methodology for expressing biological knowledge for computation and communication.
Biological networks can hold a variety of different types of information, like genetic or metabolic interactions, gene, and transcriptional regulation, protein-protein interaction (PPI), or cell signaling networks and pathways.
They often form a valuable resource for hypothesis generation and further investigations, and in the course of the analyses, they are processed and enriched with additional information from experiments.
As a result further networks are generated, whether as intermediate results that should be documented in the process, as the outcome of those analyses, or as visual representations and illustrations used in reports and publications.
As a consequence, these resulting networks do not follow anymore the strict rules the source networks were subjected, therefore a more flexible format is needed to capture their content.</p>
<p>In addition, suitable solutions for transmission conflict with those for storage, or usage in applications and analyses.
Therefore, seamless conversion between those different formats becomes as important as the data itself.</p>
<div id="the-cytoscape-exchange-cx" class="section level2">
<h2><span class="header-section-number">1.1</span> The Cytoscape Exchange (CX)</h2>
<p>A possible solution as flexible transmission format is provided by the Cytoscape Exchange (CX) format.
CX is a JSON-based, aspect-oriented data structure, which means that the network is divided into several independent modules (“aspects”).
This way, every aspect of a network, meaning nodes, edges, its attributes, and visual representations can be handled individually without interference.
Each aspect has its own schema for its contained information, and links between aspects are realized by referencing the unique internal IDs of other aspects.
The CX data model was developed by the NDEx project, in collaboration with the Cytoscape Consortium (<a href="http://www.cytoscapeconsortium.org/" class="uri">http://www.cytoscapeconsortium.org/</a>) as a transmission format between their tools, and since adopted by many others.
More details about the CX data model can be found on its documentation website: <a href="https://home.ndexbio.org/data-model/" class="uri">https://home.ndexbio.org/data-model/</a></p>
</div>
<div id="rcx---an-adaption-of-the-cx-format" class="section level2">
<h2><span class="header-section-number">1.2</span> RCX - an adaption of the CX format</h2>
<p>CX is a JSON-based data structure designed as a flexible model for transmitting networks with a focus on flexibility, modularity, and extensibility.
Although those features are widely used in common REST protocols they don’t quite fit the R way of thinking about data.</p>
<p>The RCX package provides an adaption of the CX format to standard R data formats and types to create and modify, load, export, and visualize those networks.</p>
<p>For a complete documentation see the package vignette and reference manual at Bioconductor:</p>
<p><a href="https://bioconductor.org/packages/RCX" class="uri">https://bioconductor.org/packages/RCX</a></p>
</div>
</div>
<div id="the-basics" class="section level1">
<h1><span class="header-section-number">2</span> The basics</h1>
<p>In the following, it will be explained, how to read and write networks from/to CX files, create and modify RCX networks, validate its contents and finally visualize them.</p>
<div id="read-and-write-cx-files" class="section level2">
<h2><span class="header-section-number">2.1</span> Read and write CX files</h2>
<p>Networks can be downloaded from the <a href="http://ndexbio.org/">NDEx plattform</a> as CX files in JSON format.
Those files can be read, and are automatically transformed into RCX networks that can be used in R.
Here we load a provided example network from file:</p>
<pre class="r"><code>cxFile &lt;- system.file(
  &quot;extdata&quot;, 
  &quot;Imatinib-Inhibition-of-BCR-ABL-66a902f5-2022-11e9-bb6a-0ac135e8bacf.cx&quot;, 
  package = &quot;RCX&quot;
)

rcx = readCX(cxFile)</code></pre>
<p>This network also can be accessed and downloaded from NDEx at <a href="https://www.ndexbio.org/viewer/networks/66a902f5-2022-11e9-bb6a-0ac135e8bacf" class="uri">https://www.ndexbio.org/viewer/networks/66a902f5-2022-11e9-bb6a-0ac135e8bacf</a></p>
<p>RCX networks can be saved in a similar manner:</p>
<pre class="r"><code>writeCX(rcx, &quot;path/to/some-file.cx&quot;)</code></pre>
<p>However, there might some errors occur while reading CX file.
This might happen, because the definition of the CX has changed over time, and so the definition of some aspects.
Therefore it is possible, that there are still some networks stored at the NDEx platform following a deprecated format.
In those cases it might be helpful to process the CX network step by step:</p>
<p><strong>1.</strong> just read the JSON content without parsing</p>
<pre class="r"><code>json &lt;- readJSON(cxFile)

substr(json, 1, 77)</code></pre>
<pre><code>## [{&quot;numberVerification&quot;:[{&quot;longNumber&quot;:281474976710655}]},{&quot;metaData&quot;:[{&quot;name&quot;</code></pre>
<p>This also allows to handle a CX network in JSON format, even if it comes from a different source instead of a file.</p>
<p><strong>2.</strong> parse the JSON</p>
<pre class="r"><code>aspectList &lt;- parseJSON(json)

str(aspectList, 2)</code></pre>
<pre><code>## List of 12
##  $ :List of 1
##   ..$ numberVerification:List of 1
##  $ :List of 1
##   ..$ metaData:List of 9
##  $ :List of 1
##   ..$ provenanceHistory:List of 1
##  $ :List of 1
##   ..$ nodes:List of 75
##  $ :List of 1
##   ..$ edges:List of 159
##  $ :List of 1
##   ..$ networkAttributes:List of 10
##  $ :List of 1
##   ..$ nodeAttributes:List of 1129
##  $ :List of 1
##   ..$ edgeAttributes:List of 229
##  $ :List of 1
##   ..$ cartesianLayout:List of 75
##  $ :List of 1
##   ..$ cyVisualProperties:List of 3
##  $ :List of 1
##   ..$ cyHiddenAttributes:List of 1
##  $ :List of 1
##   ..$ status:List of 1</code></pre>
<p>The result of the parsing are nested lists containing all aspects and its contents.
This format not easy to handle in R, but allows error correction previous to forming aspect and RCX objects.</p>
<p><strong>3.</strong> process the aspect data</p>
<pre class="r"><code>rcx &lt;- processCX(aspectList)</code></pre>
<p>All the above function for processing CX networks come with an option to show the performed steps.
This might be helpful for finding occurring errors:</p>
<pre class="r"><code>rcx &lt;- readCX(cxFile, verbose = TRUE)</code></pre>
<pre><code>## Read file &quot;/home/flockowak/R/x86_64-pc-linux-gnu-library/4.0/RCX/extdata/Imatinib-Inhibition-of-BCR-ABL-66a902f5-2022-11e9-bb6a-0ac135e8bacf.cx&quot;...done!
## Parse json...done!
## Parsing nodes...create aspect...done!
## Create RCX from parsed nodes...done!
## Parsing edges...create aspect...done!
## Add aspect &quot;edges&quot; to RCX...done!
## Parsing node attributes...create aspect...done!
## Add aspect &quot;nodeAttributes&quot; to RCX...done!
## Parsing edge attributes...create aspect...done!
## Add aspect &quot;edgeAttributes&quot; to RCX...done!
## Parsing network attributes...create aspect...done!
## Add aspect &quot;networkAttributes&quot; to RCX...done!
## Parsing cartesian layout...create aspect...done!
## Add aspect &quot;cartesianLayout&quot; to RCX...done!
## Parsing Cytoscape visual properties...done!
## - Create sub-objects...done!
## - Create aspect...done!
## Add aspect &quot;cyVisualProperties&quot; to RCX...done!
## Parsing Cytoscape hidden attributes...create aspect...done!
## Add aspect &quot;cyHiddenAttributes&quot; to RCX...done!
## Parsing meta-data...done!
## Ignore &quot;numberVerification&quot; aspect, not necessary in RCX!
## Can&#39;t process aspect &quot;numberVerification&quot;, so skip it...done!
## Don&#39;t know what to do with a &quot;provenanceHistory&quot; aspect!
## Can&#39;t process aspect &quot;provenanceHistory&quot;, so skip it...done!
## Ignore &quot;status&quot; aspect, not necessary in RCX!
## Can&#39;t process aspect &quot;status&quot;, so skip it...done!</code></pre>
<p>This shows, that some aspects that are contained in the CX file are ignored while creating the RCX network.
Those are for example aspects needed for transmission of the CX (<code>status</code>, <code>numberVerification</code>), or deprecated aspects (<code>provenanceHistory</code>).
For more details about the differences in aspects see <a href="Appendix_The_RCX_and_CX_Data_Model.Rmd">Appendix: The RCX and CX Data Model</a>.</p>
</div>
<div id="explore-the-rcx-object" class="section level2">
<h2><span class="header-section-number">2.2</span> Explore the RCX object</h2>
<p>The simplest way to have a look at the content of an RCX object is by printing it:</p>
<pre class="r"><code>print(rcx)
## OR:
rcx</code></pre>
<p>However, especially for large networks this can produce long and hardly readable output.
To get a better overview of the contained aspects, the mandatory and automatically generated <code>metaData</code> aspect provides information about the contained aspects.
This includes information about the number of elements or the highest used ID, if an aspect uses internal IDs:</p>
<pre class="r"><code>print(rcx$metaData)</code></pre>
<pre><code>## Meta-data:</code></pre>
<div id="htmlwidget-d7981d890d8c8ca21f97" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-d7981d890d8c8ca21f97">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8"],["nodes","edges","nodeAttributes","edgeAttributes","networkAttributes","cartesianLayout","cyVisualProperties","cyHiddenAttributes"],["1.0","1.0","1.0","1.0","1.0","1.0","1.0","1.0"],[11551,11554,null,null,null,null,null,null],[75,159,1129,229,10,75,3,1],[1,1,1,1,1,1,1,1]],"container":"<div class='horizontal-scroll'><table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>version<\/th>\n      <th>idCounter<\/th>\n      <th>elementCount<\/th>\n      <th>consistencyGroup<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>Besides exploring the RCX-object manually, a summary of the object, or single aspects, can provide more insight on them:</p>
<pre class="r"><code>summary(rcx$nodeAttributes)</code></pre>
<pre><code>##       propertyOf        name               value    
##  Total     : 1129   Length:1129        Boolean:185  
##  Unique ids:   75   Unique:34          Double : 92  
##  Min.      :11321   Class :character   Integer: 68  
##  Max.      :11551                      String :784</code></pre>
<p>We already can quickly see that there are many different node attributes are used.
The different node attributes are:</p>
<pre class="r"><code>unique(rcx$nodeAttributes$name)</code></pre>
<pre><code>##  [1] &quot;sbo&quot;                   &quot;metaId&quot;                &quot;compartmentCode&quot;       &quot;sbml type&quot;            
##  [5] &quot;sbml id&quot;               &quot;reversible&quot;            &quot;sbml compartment&quot;      &quot;cyId&quot;                 
##  [9] &quot;label&quot;                 &quot;sbml type ext&quot;         &quot;fast&quot;                  &quot;constant&quot;             
## [13] &quot;units&quot;                 &quot;boundaryCondition&quot;     &quot;derivedUnits&quot;          &quot;sbml initial amount&quot;  
## [17] &quot;value&quot;                 &quot;substanceUnits&quot;        &quot;hasOnlySubstanceUnits&quot; &quot;uniprot&quot;              
## [21] &quot;inchikey&quot;              &quot;chebi&quot;                 &quot;biocyc&quot;                &quot;cas&quot;                  
## [25] &quot;chemspider&quot;            &quot;kegg.compound&quot;         &quot;inchi&quot;                 &quot;pubchem.compound&quot;     
## [29] &quot;hmdb&quot;                  &quot;scale&quot;                 &quot;exponent&quot;              &quot;kind&quot;                 
## [33] &quot;multiplier&quot;            &quot;unitSid&quot;</code></pre>
</div>
<div id="visualize-the-network" class="section level2">
<h2><span class="header-section-number">2.3</span> Visualize the network</h2>
<p>This package provides simple functions to visualize the network encoded in the RCX object.</p>
<pre class="r"><code>visualize(rcx)</code></pre>
<div class="figure">
<img src="/home/flockowak/Projekte/Workflow/Workflow/img/RCX/1_network_orig.png" alt="" />
<p class="caption">Imatinib Inhibition of BCR-ABL (66a902f5-2022-11e9-bb6a-0ac135e8bacf)</p>
</div>
<p>The visualization also utilizes the same JavaScript library as the NDEx platform.
Therefore the visual result is the same as when the network is uploaded to the NDEx platform.
Additionally, this allows the visualization to be exported as a single HTML file, which can directly be hosted on a web-server or included in existing websites:</p>
<pre class="r"><code>writeHTML(rcx, &quot;path/to/some-file.html&quot;)</code></pre>
<p>Networks with many nodes and edges, or those without a provided layout, often are difficult to interpret visually.
To untangle the “hairball” it is possible to apply different layout option provided by the <a href="https://js.cytoscape.org/">Cytoscape.js</a> framework.
A force driven layout is a good starting point in those cases.
For demonstration purposes, let’s delete the visual layout of our network first:</p>
<pre class="r"><code>## save them for later
originalVisualProperties &lt;- rcx$cyVisualProperties

## and delete them from the RCX network
rcx$cyVisualProperties &lt;- NULL
rcx &lt;- updateMetaData(rcx)

print(rcx$metaData)</code></pre>
<pre><code>## Meta-data:</code></pre>
<div id="htmlwidget-3d41af86ae7b2f81c2dd" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-3d41af86ae7b2f81c2dd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7"],["nodes","edges","nodeAttributes","edgeAttributes","networkAttributes","cartesianLayout","cyHiddenAttributes"],["1.0","1.0","1.0","1.0","1.0","1.0","1.0"],[11551,11554,null,null,null,null,null],[75,159,1129,229,10,75,1],[1,1,1,1,1,1,1]],"container":"<div class='horizontal-scroll'><table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>version<\/th>\n      <th>idCounter<\/th>\n      <th>elementCount<\/th>\n      <th>consistencyGroup<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>Lets have a look at the visualization now:</p>
<pre class="r"><code>visualize(rcx, layout = c(name = &quot;cose&quot;))</code></pre>
<div class="figure">
<img src="/home/flockowak/Projekte/Workflow/Workflow/img/RCX/1_network_no_cvp.png" alt="" />
<p class="caption">Force-driven visualization of the network without visual properties</p>
</div>
<p>Unfortunately no labels are shown in the network.
We can fix that by defining a simple pass-through mapping for the node labels and add it to the network:</p>
<pre class="r"><code>cyMapping &lt;- createCyVisualPropertyMappings(
  name= &quot;NODE_LABEL&quot; ,
  type = &quot;PASSTHROUGH&quot;,
  definition = &quot;COL=label,T=string&quot;
)

cyVisualProperties &lt;- createCyVisualProperties(
  defaultNodes = createCyVisualProperty(
    mappings = cyMapping
  )
)

rcx &lt;- updateCyVisualProperties(rcx, cyVisualProperties)</code></pre>
<p>By default the visualization is opened in RStudio, but it also can be forced to open in an external browser.
The <code>Cytoscape.js</code> parameters used to customize the layout algorithm can be found in it documentation at <a href="https://js.cytoscape.org/#layouts/cose" class="uri">https://js.cytoscape.org/#layouts/cose</a></p>
<pre class="r"><code>visualize(
  rcx, 
  layout = c(
    name=&quot;cose&quot;, 
    idealEdgeLength=&quot;80&quot;, 
    edgeElasticity=&quot;250&quot;), 
  openExternal = TRUE)</code></pre>
<p><img src="/home/flockowak/Projekte/Workflow/Workflow/img/RCX/1_network_cose_with_label.png" /></p>
</div>
<div id="validation" class="section level2">
<h2><span class="header-section-number">2.4</span> Validation</h2>
<p>The correctness of the RCX network is important for the conversion to CX, and therefore to be used at all platforms and tools.
The validity of the an RCX network can be checked simply:</p>
<pre class="r"><code>validate(rcx)</code></pre>
<pre><code>## Checking Nodes Aspect:
## - Is object of class &quot;NodesAspect&quot;...OK
## - All required columns present (id)...OK
## - Column (id) doesn&#39;t contain any NA values...OK
## - At least one node present...OK
## - Column (id) contains only unique values...OK
## - Column (id) only contains numeric values...OK
## - Column (id) only contains positive (&gt;=0) values...OK
## - No merge artefacts present (i.e. column with old ids: oldId)...OK
## - Only allowed columns present (id, name, represents)...OK
## &gt;&gt; Nodes Aspect: OK
## 
## Checking Edges Aspect:
## - Is object of class &quot;EdgesAspect&quot;...OK
## - All required columns present (id, source, target)...OK
## - Column (id) doesn&#39;t contain any NA values...OK
## - Column (id) contains only unique values...OK
## - Column (id) only contains numeric values...OK
## - Column (id) only contains positive (&gt;=0) values...OK
## - Column (source) doesn&#39;t contain any NA values...OK
## - Column (source) only contains numeric values...OK
## - Column (source) only contains positive (&gt;=0) values...OK
## - Column (target) doesn&#39;t contain any NA values...OK
## - Column (target) only contains numeric values...OK
## - Column (target) only contains positive (&gt;=0) values...OK
## - No merge artefacts present (i.e. column with old ids: oldId)...OK
## - Only allowed columns present (id, source, target, name, interaction)...OK
## &gt;&gt; Edges Aspect: OK
## 
## Checking Node Attributes Aspect:
## - Is object of class &quot;NodeAttributesAspect&quot;...OK
## - All required columns present (propertyOf, name, value, dataType, isList)...OK
## - Combination of columns (propertyOf, name) contains only unique values...OK
## - Column (propertyOf) doesn&#39;t contain any NA values...OK
## - Column (propertyOf) only contains numeric values...OK
## - Column (propertyOf) only contains positive (&gt;=0) values...OK
## - Is the column (name) a character vector...OK
## - Is the column (value) a list...OK
## - Column (dataType) doesn&#39;t contain any NA values...OK
## - Is the column (dataType) a character vector...OK
## - All values of dataType are in the allowed set (boolean, integer, long, double, string)...OK
## - Column (isList) doesn&#39;t contain any NA values...OK
## - Column (isList) only contains logical values...OK
## - Only allowed columns present (propertyOf, name, value, dataType, isList)...OK
## &gt;&gt; Node Attributes Aspect: OK
## 
## Checking Edge Attributes Aspect:
## - Is object of class &quot;EdgeAttributesAspect&quot;...OK
## - All required columns present (propertyOf, name, value, dataType, isList)...OK
## - Combination of columns (propertyOf, name) contains only unique values...OK
## - Column (propertyOf) doesn&#39;t contain any NA values...OK
## - Column (propertyOf) only contains numeric values...OK
## - Column (propertyOf) only contains positive (&gt;=0) values...OK
## - Is the column (name) a character vector...OK
## - Is the column (value) a list...OK
## - Column (dataType) doesn&#39;t contain any NA values...OK
## - Is the column (dataType) a character vector...OK
## - All values of dataType are in the allowed set (boolean, integer, long, double, string)...OK
## - Column (isList) doesn&#39;t contain any NA values...OK
## - Column (isList) only contains logical values...OK
## - Only allowed columns present (propertyOf, name, value, dataType, isList)...OK
## &gt;&gt; Edge Attributes Aspect: OK
## 
## Checking Network Attributes Aspect:
## - Is object of class &quot;NetworkAttributesAspect&quot;...OK
## - All required columns present (name, value, dataType, isList)...OK
## - Column (name) contains only unique values...OK
## - Is the column (name) a character vector...OK
## - Is the column (value) a list...OK
## - Column (dataType) doesn&#39;t contain any NA values...OK
## - Is the column (dataType) a character vector...OK
## - All values of dataType are in the allowed set (boolean, integer, long, double, string)...OK
## - Column (isList) doesn&#39;t contain any NA values...OK
## - Column (isList) only contains logical values...OK
## - Only allowed columns present (name, value, dataType, isList)...OK
## &gt;&gt; Network Attributes Aspect: OK
## 
## Checking Cartesian Layout Aspect:
## - Is object of class &quot;CartesianLayoutAspect&quot;...OK
## - All required columns present (node, x, y)...OK
## - Column (node) contains only unique values...OK
## - Column (node) only contains numeric values...OK
## - Column (node) doesn&#39;t contain any NA values...OK
## - Column (x) only contains numeric values...OK
## - Column (x) doesn&#39;t contain any NA values...OK
## - Column (y) only contains numeric values...OK
## - Column (y) doesn&#39;t contain any NA values...OK
## - Only allowed columns present (node, x, y, z)...OK
## &gt;&gt; Cartesian Layout Aspect: OK
## 
## Checking Cytoscape Visual Properties Aspect:
## - Is object of class &quot;CyVisualPropertiesAspect&quot;...OK
## - Is object a list...OK
## - Is a named list (&quot;network&quot;, &quot;nodes&quot;, &quot;edges&quot;, &quot;defaultNodes&quot; or &quot;defaultEdges&quot;)...OK
## - The list only contains entries of class &quot;CyVisualProperty&quot;...OK
## - At least one of the elements present (network, nodes, edges, defaultNodes, defaultEdges)...OK
## For defaultNodes: Checking Cytoscape Visual Property sub-aspect:
## - Is object of class &quot;CyVisualProperty&quot;...OK
## - Is object a list...OK
## - Is a named list (&quot;appliesTo&quot;, &quot;view&quot;, &quot;properties&quot;, &quot;dependencies&quot; or &quot;mappings&quot;)...OK
## - List element (appliesTo) only contains numeric values...OK
## - List element (view) only contains numeric values...OK
## - Combination of columns (appliesTo, view) contains only unique values...OK
## - At least one of the elements present (properties, dependencies, mappings)...OK
##   Checking Cytoscape Visual Property Mappings:
## - Is mappings a list...OK
## - The list only contains entries of class &quot;CyVisualPropertyMappings&quot;...OK
## - All required columns present (name, type, definition)...OK
## - All list elements of name contain only character values...OK
## - All list elements of name don&#39;t contain any NA values...OK
## - All list elements (name) contain only unique values...OK
## - All list elements of type contain only character values...OK
## - All list elements of type don&#39;t contain any NA values...OK
## - All list elements of definition contain only character values...OK
## - All list elements of definition don&#39;t contain any NA values...OK
## - Only allowed columns present (name, type, definition)...OK
## &gt;&gt; Cytoscape Visual Property sub-aspect: OK
## &gt;&gt; Cytoscape Visual Properties Aspect: OK
## 
## Checking Cytoscape Hidden Attributes Aspect:
## - Is object of class &quot;CyHiddenAttributesAspect&quot;...OK
## - All required columns present (name, value, dataType, isList)...OK
## - Column (name) contains only unique values...OK
## - Is the column (name) a character vector...OK
## - Is the column (value) a list...OK
## - Column (dataType) doesn&#39;t contain any NA values...OK
## - Is the column (dataType) a character vector...OK
## - All values of dataType are in the allowed set (boolean, integer, long, double, string)...OK
## - Column (isList) doesn&#39;t contain any NA values...OK
## - Column (isList) only contains logical values...OK
## - Only allowed columns present (name, value, dataType, isList)...OK
## &gt;&gt; Cytoscape Hidden Attributes Aspect: OK
## 
## Checking RCX:
## - Is object of class &quot;RCX&quot;...OK
## - nodes aspect is present...OK
## - Validate nodes aspect...OK
## - Validate edges aspect...OK
##   - Reference aspect (nodes) present and correct...OK
##   - All id references exist (EdgesAspect$source ids in NodesAspect$id)...OK
##   - All id references exist (EdgesAspect$target ids in NodesAspect$id)...OK
## - Validate node attributes aspect...OK
##   - Reference aspect (nodes) present and correct...OK
##   - All id references exist (NodeAttributesAspect$propertyOf ids in NodesAspect$id)...OK
## - Validate edge attributes aspect...OK
##   - Reference aspect (edges) present and correct...OK
##   - All id references exist (EdgeAttributesAspect$propertyOf ids in EdgesAspect$id)...OK
## - Validate network attributes aspect...OK
## - Validate cartesian layout aspect...OK
##   - Reference aspect (nodes) present and correct...OK
##   - All id references exist (CartesianLayoutAspect$node ids in NodesAspect$id)...OK
## - Validate cytoscape visual property aspect...OK
## - Validate cytoscape hidden attributes aspect...OK
## &gt;&gt; RCX: OK</code></pre>
<p>The verbose output also shows at which tests are performed.
Let’s manually introduce an error in an aspect and validate only this one again.
Therefore we simply duplicate a node ID.</p>
<pre class="r"><code>nodes &lt;- rcx$nodes
nodes$id[1] &lt;- nodes$id[2]

test &lt;- validate(nodes)</code></pre>
<pre><code>## Checking Nodes Aspect:
## - Is object of class &quot;NodesAspect&quot;...OK
## - All required columns present (id)...OK
## - Column (id) doesn&#39;t contain any NA values...OK
## - At least one node present...OK
## - Column (id) contains only unique values...FAIL
## - Column (id) only contains numeric values...OK
## - Column (id) only contains positive (&gt;=0) values...OK
## - No merge artefacts present (i.e. column with old ids: oldId)...OK
## - Only allowed columns present (id, name, represents)...OK
## &gt;&gt; Nodes Aspect: FAIL</code></pre>
<pre class="r"><code>test</code></pre>
<pre><code>## [1] FALSE</code></pre>
<p>As expected, the test fails and returns <code>FALSE</code>.</p>
</div>
</div>
<div id="get-information-about-the-networks" class="section level1">
<h1><span class="header-section-number">3</span> Get information about the networks</h1>
<p>It is always useful to get some basic information about the network.
For example this can be the number of elements in the different aspects of the network:</p>
<pre class="r"><code>countElements(rcx)</code></pre>
<pre><code>##              NodesAspect           MetaDataAspect              EdgesAspect     NodeAttributesAspect 
##                       75                       NA                      159                     1129 
##     EdgeAttributesAspect  NetworkAttributesAspect    CartesianLayoutAspect CyHiddenAttributesAspect 
##                      229                       10                       75                        1 
## CyVisualPropertiesAspect 
##                        1</code></pre>
<p>This also works for single aspects:</p>
<pre class="r"><code>countElements(rcx$nodes)</code></pre>
<pre><code>## [1] 75</code></pre>
<p>To determine, if an aspect contains IDs (on the contrary to knowing it beforehand), this can be checked with:</p>
<pre class="r"><code>hasIds(rcx$nodes)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>If an aspect has IDs, one can check what the highest used ID is, to know at which ID the next elements have to continue before adding them.
This can even be done for the whole network at once.</p>
<pre class="r"><code>maxId(rcx$nodes)</code></pre>
<pre><code>## [1] 11551</code></pre>
<pre class="r"><code>maxId(rcx)</code></pre>
<pre><code>## NodesAspect EdgesAspect 
##       11551       11554</code></pre>
<p>Since we now know, that the nodes aspect has IDs, we can also determine the name of the property that holds the IDs:</p>
<pre class="r"><code>idProperty(rcx$nodes)</code></pre>
<pre><code>## [1] &quot;id&quot;</code></pre>
<p>Other aspects use those IDs to reference to them.
Let’s find out, which aspect is referred by others:</p>
<pre class="r"><code>referredBy(rcx)</code></pre>
<pre><code>## $NodesAspect
## [1] &quot;EdgesAspect&quot;           &quot;NodeAttributesAspect&quot;  &quot;CartesianLayoutAspect&quot;
## 
## $EdgesAspect
## [1] &quot;EdgeAttributesAspect&quot;</code></pre>
<p>The nodes aspect is referred by the edges aspect, so we can find out which properties of the edges aspect refer to it:</p>
<pre class="r"><code>refersTo(rcx$edges)</code></pre>
<pre><code>##        source        target 
## &quot;NodesAspect&quot; &quot;NodesAspect&quot;</code></pre>
<p>It might have gotten to your attention, that there is a difference between the aspect name and the aspect class.
This has been done intentionally to avoid naming conflicts.
The different naming can be converted to each other:</p>
<pre class="r"><code>## all classes
aspectClasses</code></pre>
<pre><code>##                        rcx                   metaData                      nodes                      edges 
##                      &quot;RCX&quot;           &quot;MetaDataAspect&quot;              &quot;NodesAspect&quot;              &quot;EdgesAspect&quot; 
##             nodeAttributes             edgeAttributes          networkAttributes            cartesianLayout 
##     &quot;NodeAttributesAspect&quot;     &quot;EdgeAttributesAspect&quot;  &quot;NetworkAttributesAspect&quot;    &quot;CartesianLayoutAspect&quot; 
##                   cyGroups         cyVisualProperties         cyHiddenAttributes         cyNetworkRelations 
##           &quot;CyGroupsAspect&quot; &quot;CyVisualPropertiesAspect&quot; &quot;CyHiddenAttributesAspect&quot; &quot;CyNetworkRelationsAspect&quot; 
##              cySubNetworks              cyTableColumn 
##      &quot;CySubNetworksAspect&quot;      &quot;CyTableColumnAspect&quot;</code></pre>
<pre class="r"><code>## class of nodes
aspectName2Class(&quot;nodes&quot;)</code></pre>
<pre><code>##         nodes 
## &quot;NodesAspect&quot;</code></pre>
<pre class="r"><code>## accession name of NodesAspect
aspectClass2Name(&quot;NodesAspect&quot;)</code></pre>
<pre><code>## [1] &quot;nodes&quot;</code></pre>
<pre class="r"><code>## back and forth
class(rcx[[aspectClass2Name(&quot;NodesAspect&quot;)]])</code></pre>
<pre><code>## [1] &quot;NodesAspect&quot; &quot;data.frame&quot;</code></pre>
</div>
<div id="conversion-to-r-graph-data-models" class="section level1">
<h1><span class="header-section-number">4</span> Conversion to R graph data models</h1>
<p>The RCX package provides conversion to and from objects of iGraph and Bioconductor graph, both widely used libraries for graph manipulation and network analysis.</p>
<div id="igraph" class="section level2">
<h2><span class="header-section-number">4.1</span> igraph</h2>
<p>igraph is a library for creating and manipulating graphs and analyzing (especially large) networks.
RCX networks can be simply converted to igraph objects as follows:</p>
<pre class="r"><code>ig &lt;- toIgraph(rcx)
summary(ig)</code></pre>
<pre><code>## IGRAPH e55cc54 UN-- 75 159 -- 
## + attr: attribute...name (g/c), description (g/c), smpdb (g/c), sbmlVersion (g/c), sbml id (g/c),
## | metaId (g/c), sbmlNetwork (g/c), label (g/c), networkType (g/c), reference (g/c), name (v/n), id
## | (v/n), sbo (v/c), metaId (v/c), compartmentCode (v/c), compartmentCode...dataType (v/c), sbml type
## | (v/c), sbml id (v/c), reversible (v/c), sbml compartment (v/c), cyId (v/c), label (v/c), sbml type
## | ext (v/c), fast (v/c), constant (v/c), units (v/c), boundaryCondition (v/c), derivedUnits (v/c),
## | sbml initial amount (v/c), value (v/c), substanceUnits (v/c), hasOnlySubstanceUnits (v/c), uniprot
## | (v/c), inchikey (v/c), chebi (v/c), biocyc (v/c), cas (v/c), chemspider (v/c), kegg.compound (v/c),
## | inchi (v/c), pubchem.compound (v/c), hmdb (v/c), scale (v/c), scale...dataType (v/c), exponent
## | (v/c), kind (v/c), multiplier (v/c), unitSid (v/c), cartesianLayout...x (v/n), cartesianLayout...y
## | (v/n), id (e/n), interaction (e/c), sbo (e/c), stoichiometry (e/c), interaction type (e/c)</code></pre>
<p>To ensure a consistent conversion in both direction, some conventions have to be matched:
For example avoiding collisions between <code>name</code> used in nodes, <code>name</code> used as nodes attribute, and the <code>name</code> used by igraph, the first two were re-named in igraph to <code>nodeName</code> and <code>attribute...name</code> respectively.
A similar convention is used for the cartesian coordinates:
In the igraph object they are saved as <code>cartesianLayout...x</code>, <code>cartesianLayout...y</code>, and <code>cartesianLayout...z</code>.
If the vertex attributes match this naming convention, they assigned to the correct aspect when converted back:</p>
<pre class="r"><code>rcxFromIg &lt;- fromIgraph(ig)</code></pre>
<p>Igraph objects can not hold information about the visual representation of the network, and therefore the Cytoscape aspects are lost in the conversion.
But we can simply add the <code>CyVisualProperties</code> aspect we saved previously to get back to the original layout:</p>
<pre class="r"><code>rcxFromIg &lt;- updateCyVisualProperties(rcxFromIg, originalVisualProperties)</code></pre>
<pre class="r"><code>visualize(rcxFromIg)</code></pre>
</div>
</div>
<div id="bioconductor-graph-graphnel" class="section level1">
<h1><span class="header-section-number">5</span> Bioconductor graph (graphNEL)</h1>
<p>Bioconductor graph is a package that implements some simple graph handling capabilities.
It can handle multi-edges, but only if the graph is directed and the source and target start and end not between the same nodes.
Unfortunaltelly this is the case in our sample network.
A quick fix is simply switching the direction of source and target for the multi-edges:</p>
<pre class="r"><code>dubEdges &lt;- duplicated(rcx$edges[c(&quot;source&quot;, &quot;target&quot;)])

s &lt;- rcx$edges$source
rcx$edges$source[dubEdges] &lt;- rcx$edges$target[dubEdges]
rcx$edges$target[dubEdges] &lt;- s[dubEdges]

gNel &lt;- toGraphNEL(rcx, directed = TRUE)
rcxBack &lt;- fromGraphNEL(gNel)</code></pre>
<p>Then we can simply convert the RCX to a graphNEL object:</p>
<pre class="r"><code>gNel &lt;- toGraphNEL(rcx, directed = TRUE)
gNel</code></pre>
<pre><code>## A graphNEL graph with directed edges
## Number of Nodes = 75 
## Number of Edges = 159</code></pre>
<p>For the conversion igraph is used as an intermediate format, therefore, the same conventions apply as for the igraph conversion.
The conversion back to the RCX object works analogously:</p>
<pre class="r"><code>rcxBack &lt;- fromGraphNEL(gNel)
print(rcxBack$metaData)</code></pre>
<pre><code>## Meta-data:</code></pre>
<div id="htmlwidget-0e87cc7af71c4c3f6819" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0e87cc7af71c4c3f6819">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["nodes","edges","nodeAttributes","edgeAttributes","networkAttributes","cartesianLayout"],["1.0","1.0","1.0","1.0","1.0","1.0"],[11551,11554,null,null,null,null],[75,159,1129,388,10,75],[1,1,1,1,1,1]],"container":"<div class='horizontal-scroll'><table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>version<\/th>\n      <th>idCounter<\/th>\n      <th>elementCount<\/th>\n      <th>consistencyGroup<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","columnDefs":[{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>As igraph, graphNEL objects can not hold information about the visual representation of the network, so here too we can restore the original layout by adding the <code>CyVisualProperties</code> aspect we saved previously.</p>
<div class="pagination">
<div class="prev">
Previous: <a href="Ndexr.html">ndexr</a>
</div>
<div class="curr">
RCX
</div>
<div class="next">
Next: <a href="DataExplorationAndPreparation.html">Data Exploration and Preparation</a>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">6</span> Session info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] XML_3.99-0.8        httr_1.4.2          RJSONIO_1.3-1.6     pacman_0.5.1        devtools_2.4.2     
##  [6] usethis_2.1.2       timeSeries_3062.100 timeDate_3043.102   pander_0.6.4        xtable_1.8-4       
## [11] stringr_1.4.0       BiocStyle_2.18.1    RCX_0.99.1          ndexr_1.17.0        igraph_1.2.7       
## [16] gplots_3.1.1        dplyr_1.0.7         RColorBrewer_1.1-2  survival_3.2-7     
## 
## loaded via a namespace (and not attached):
##  [1] pkgload_1.2.3       jsonlite_1.7.2      splines_4.0.3       gtools_3.9.2        assertthat_0.2.1   
##  [6] BiocManager_1.30.16 highr_0.9           stats4_4.0.3        remotes_2.4.1       yaml_2.2.1         
## [11] sessioninfo_1.1.1   pillar_1.6.4        lattice_0.20-41     glue_1.4.2          digest_0.6.28      
## [16] htmltools_0.5.2     Matrix_1.2-18       plyr_1.8.6          pkgconfig_2.0.3     bookdown_0.24      
## [21] purrr_0.3.4         webshot_0.5.2       processx_3.5.2      tibble_3.1.5        generics_0.1.1     
## [26] ellipsis_0.3.2      DT_0.21             withr_2.4.2         cachem_1.0.6        BiocGenerics_0.36.1
## [31] cli_3.0.1           mime_0.12           magrittr_2.0.1      crayon_1.4.1        ps_1.6.0           
## [36] memoise_2.0.0       evaluate_0.14       fs_1.5.0            fansi_0.5.0         pkgbuild_1.2.0     
## [41] graph_1.68.0        prettyunits_1.1.1   tools_4.0.3         formatR_1.11        lifecycle_1.0.1    
## [46] callr_3.7.0         compiler_4.0.3      jquerylib_0.1.4     caTools_1.18.2      tinytex_0.34       
## [51] rlang_0.4.12        grid_4.0.3          htmlwidgets_1.5.4   crosstalk_1.1.1     bitops_1.0-7       
## [56] rmarkdown_2.11      testthat_3.1.0      DBI_1.1.1           curl_4.3.2          markdown_1.1       
## [61] R6_2.5.1            knitr_1.36          fastmap_1.1.0       utf8_1.2.2          rprojroot_2.0.2    
## [66] desc_1.4.0          KernSmooth_2.23-17  stringi_1.7.5       parallel_4.0.3      Rcpp_1.0.7         
## [71] vctrs_0.3.8         tidyselect_1.1.1    xfun_0.27</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
