<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-03-30" />

<title>Explore different ways of visualizing the networks</title>

<script src="Visualization_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Visualization_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="Visualization_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Visualization_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Visualization_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Visualization_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="Visualization_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="Visualization_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="Visualization_files/navigation-1.1/tabsets.js"></script>
<link href="Visualization_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Visualization_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="/home/flockowak/R/x86_64-pc-linux-gnu-library/4.0/BiocStyle/resources/html/bioconductor.css" type="text/css" />
<link rel="stylesheet" href="custom.css" type="text/css" />
<link rel="stylesheet" href="bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {

}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 246px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">Explore different ways of visualizing the networks</h1>
<p class="author-name">Florian J. Auer<span class="affil-mark">1</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>IT-Infrastructure for Translational Medical Research, University of Augsburg</p>
<h4 class="date">03/30/2022</h4>

</div>


<ul>
<li><a href="Overview.html">Overview</a> <em>A short introduction</em>
<ul>
<li><a href="SoftwareRequirements.html">Software requirements</a> <em>Installation of required packages</em></li>
<li><a href="Ndexr.html">ndexr</a> <em>ndexr - interface with the network data exchange (NDEx)</em></li>
<li><a href="RCX.html">RCX</a> <em>RCX - R implementation of the Cytoscape Exchange (CX) format</em></li>
</ul></li>
<li><a href="DataExplorationAndPreparation.html">Data Exploration and Preparation</a> <em>Loading the source data, explore the content and prepare the data for its usage</em></li>
<li><a href="DataIntegration.html">Data Integration</a> <em>Combine the network data with patient information, gene expression and relevance score</em></li>
<li><a href="Visualization.html">Visualization</a> <em>Explore different ways of visualizing the networks</em>
<ul>
<li><a href="MetaRelSubNetVis.html">MetaRelSubNetVis</a> <em>Comparative web-based visualization</em></li>
<li><a href="NDExEdit.html">NDExEdit</a> <em>Web-based data-dependant visualization with NDExEdit</em></li>
<li><a href="RCXvis.html">RCX based</a> <em>R-based visualization with the RCX package</em></li>
<li><a href="CytoscapeAndRCy3.html">Cytoscape and RCy3</a> <em>Using the RCy3 package to remotly control Cytoscape</em></li>
</ul></li>
</ul>
<div id="overview" class="section level1">
<h1><span class="header-section-number">1</span> Overview</h1>
<p>The network is also available on the NDEx platform as “Combined patient-specific breast cancer subnetworks”:</p>
<p><a href="https://www.ndexbio.org/viewer/networks/079f4c66-3b77-11ec-b3be-0ac135e8bacf" class="uri">https://www.ndexbio.org/viewer/networks/079f4c66-3b77-11ec-b3be-0ac135e8bacf</a></p>
<p>The network does not have any visualization yet.
By default a circular layout is applied to at the NDEx platform to be able to display the network in a way:</p>
<div class="figure">
<img src="img/Combined-subnetworks-on-NDEx.png" alt="" />
<p class="caption">Combined subnetworks on the NDEx platform with default circular layout)</p>
</div>
<p>In the following chapters different tools will be used to visualize the data and document the performed steps in a reproducible manner.</p>
</div>
<div id="metarelsubnetvis" class="section level1">
<h1><span class="header-section-number">2</span> <a href="MetaRelSubNetVis.html">MetaRelSubNetVis</a></h1>
<p>The web application MetaRelSubNetVis is able to work on the combined network without any preset visualization.
It used the network stored on the NDEx platform to generate linkable and reproducible network visualizations that can be included in other applications.
It visualizes either single patients or comparative visualizations of one patient from each group, namely metastatic and non-metastatic patients.</p>
<div style="height: 50em;">
<iframe src="https://frankkramer-lab.github.io/MetaRelSubNetVis?uuid=a420aaee-4be9-11ec-b3be-0ac135e8bacf&amp;pa=GSM615368&amp;pb=GSM615184&amp;th=0.00029828155&amp;sel=1106,3406&amp;col=1&amp;size=1&amp;all=false&amp;shared=false&amp;mtb=true&amp;sb=2" allowfullscreen allow="autoplay" data-external="1" style="-webkit-transform:scale(0.5);-moz-transform-scale(0.5);-moz-transform-origin: top left; -webkit-transform-origin: top left; -o-transform-origin: top left; -ms-transform-origin: top left; transform-origin: top left;width: 200%;height: 100em;">
</iframe>
</div>
<p><a href="https://frankkramer-lab.github.io/MetaRelSubNetVis?uuid=a420aaee-4be9-11ec-b3be-0ac135e8bacf&amp;pa=GSM615368&amp;pb=GSM615184&amp;th=0.00029828155&amp;sel=1106,3406&amp;col=1&amp;size=1&amp;all=false&amp;shared=false&amp;mtb=true&amp;sb=2" class="uri">https://frankkramer-lab.github.io/MetaRelSubNetVis?uuid=a420aaee-4be9-11ec-b3be-0ac135e8bacf&amp;pa=GSM615368&amp;pb=GSM615184&amp;th=0.00029828155&amp;sel=1106,3406&amp;col=1&amp;size=1&amp;all=false&amp;shared=false&amp;mtb=true&amp;sb=2</a></p>
</div>
<div id="ndexedit" class="section level1">
<h1><span class="header-section-number">3</span> <a href="NDExEdit.html">NDExEdit</a></h1>
<p>With NDExEdit networks can be loaded from the NDEx platform and visual styles created or modified.
The network data and its distribution can be explored within the web application, and based on this data custom visual properties adjusted.
The created network then can be exported directly back to the NDEx platform or downloaded as file.</p>
</div>
<div id="rcx-based" class="section level1">
<h1><span class="header-section-number">4</span> <a href="RCXvis.html">RCX based</a></h1>
<p>The RCX package was already used for data preparation and integration, but it also can be used for creating visualizations from within R without the need of additional software.
In this chapter the details of how the visual properties can be created and even visualized with the RCX package.</p>
<pre class="r"><code>visualize(combined_subnetwork_rcx, c(name = &quot;circle&quot;))</code></pre>
<div class="figure">
<img src="img/Combined_subnetworks.png" alt="" />
<p class="caption">Combined subnetworks with circle (circular) layout (default on NDEx)</p>
</div>
</div>
<div id="cytoscape-and-rcy3" class="section level1">
<h1><span class="header-section-number">5</span> <a href="CytoscapeAndRCy3.html">Cytoscape and RCy3</a></h1>
<p>Cytoscape is one of the most prominent tools for the visualization of biological networks.
It is a java based software that is primarily designed for the interactive application of visual styles.
However, with the package RCy3 Cytoscape can be controlled remotely from within R.
This package can be used to achieve the same visual representation as the tools before.</p>
<div class="pagination">
<div class="prev">
Previous: <a href="DataIntegration.html">Data Integration</a>
</div>
<div class="curr">
Visualization
</div>
<div class="next">
Next: <a href="MetaRelSubNetVis.html">MetaRelSubNetVis</a>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
